### PR TITLE
Bound region range by query range

### PR DESCRIPTION
**Pull request by ekexium**
In `store_stream_for_range`, we passed the `region_range` to following steps and use it in request, which is incorrect if `region_range` is not a subset of origianl query range. We should take the common part of the two.

### CODE REVIEW

File: src/request.rs [hunk 1]
```diff
@@ -188,16 +188,30 @@ pub fn store_stream_for_range<PdC: PdClient>(
...
-            (range, store)
+            let region_range = store.region.range();
+            (bound_range(region_range, range.clone()), store)
         })
         .into_stream()
         .boxed()
 }
 
+/// The range used for request should be the common range of `region_range` and `range`.
+fn bound_range(region_range: (Key, Key), range: BoundRange) -> (Key, Key) {
```
Comment (nrc) [inline src/request.rs:201]: A better name might be `intersection`.

File: src/request.rs [hunk 2]
```diff
@@ -188,16 +188,30 @@ pub fn store_stream_for_range<PdC: PdClient>(
...
+    let (mut lower, mut upper) = region_range;
+    let (lower_bound, upper_bound) = range.into_keys();
+    if lower_bound > lower {
+        lower = lower_bound;
+    }
+    if let Some(up) = upper_bound {
+        if up < upper || upper.is_empty() {
```
Comment (nrc) [inline src/request.rs:208]: Why do we need the `is_empty` special case?

File: src/request.rs [hunk 3]
```diff
@@ -188,16 +188,30 @@ pub fn store_stream_for_range<PdC: PdClient>(
...
+            upper = up;
+        }
+    }
+    (lower, upper)
```
Comment (nrc) [inline src/request.rs:212]: I think you can use min, max, and unwrap_or, to avoid the mutation and conditionals

File: src/request.rs (hunk 2 above)
Comment (ekexium) [reply src/request.rs:208]: If `upper` is empty, which means it's upper unbounded, `up < upper` does not hold. We still need to bound it.

File: src/request.rs [hunk 4]
```diff
@@ -188,16 +192,28 @@ pub fn store_stream_for_range<PdC: PdClient>(
...
         })
         .into_stream()
         .boxed()
 }
 
+/// The range used for request should be the intersection of `region_range` and `range`.
+fn bound_range(region_range: (Key, Key), range: BoundRange) -> (Key, Key) {
+    let (lower, upper) = region_range;
+    let (lower_bound, upper_bound) = range.into_keys();
+    let upper_bound = upper_bound.unwrap_or_default();
```
Comment (nrc) [inline src/request.rs:208]: I think you can match on upper_bound, without doing this unwrap

**Review by nrc — COMMENTED** [review]
looks much better!

**Review by nrc — APPROVED** [review]
lgtm

**Pull request merged**